### PR TITLE
Add diagram links for StrokeJoin and StrokeCap

### PR DESCRIPTION
**Pull request by gspencergoog**
This updates the documentation for StrokeJoin and StrokeCap to contain animated diagrams that describe their function. The actual diagrams reside in the assets-for-api-docs repo.

### CODE REVIEW

File: lib/ui/painting.dart [hunk 1]
```diff
@@ -824,37 +824,73 @@ enum FilterQuality {
 
 /// Styles to use for line endings.
 ///
-/// See [Paint.strokeCap].
+/// See also:
+///
+/// * [Paint.strokeCap] for how this value is used.
+/// * [StrokeJoin] for the different kinds of line segment joins.
```
Comment (Hixie) [inline lib/ui/painting.dart]: indent the lists by one

File: lib/ui/painting.dart [hunk 2]
```diff
@@ -824,37 +824,73 @@ enum FilterQuality {
...
 // These enum values must be kept in sync with SkPaint::Cap.
 enum StrokeCap {
   /// Begin and end contours with a flat edge and no extension.
+  ///
+  /// ![Butt Cap Diagram](https://flutter.github.io/assets-for-api-docs/assets/dart-ui/butt_cap.png)
```
Comment (Hixie) [inline lib/ui/painting.dart]: "Butt Cap Diagram" is worthless alternative text. (It's a caption, not alternative text.) See https://html.spec.whatwg.org/multipage/images.html#alt

File: lib/ui/painting.dart [hunk 3]
```diff
@@ -824,37 +824,73 @@ enum FilterQuality {
...
+  /// The cap is zero-width here, so there is nothing to highlight. Compare to
+  /// the [square] cap, which does have a width.
   butt,
 
   /// Begin and end contours with a semi-circle extension.
+  ///
+  /// ![Round Cap Diagram](https://flutter.github.io/assets-for-api-docs/assets/dart-ui/round_cap.png)
+  ///
+  /// (The cap is colored here to highlight it: in normal use it is the same
+  /// color as the line)
```
Comment (Hixie) [inline lib/ui/painting.dart]: "here" doesn't make sense for people who can't see the image.

File: lib/ui/painting.dart [hunk 4]
```diff
@@ -824,37 +824,73 @@ enum FilterQuality {
...
+///
+/// * [Paint.strokeCap] for how this value is used.
+/// * [StrokeJoin] for the different kinds of line segment joins.
 // These enum values must be kept in sync with SkPaint::Cap.
 enum StrokeCap {
   /// Begin and end contours with a flat edge and no extension.
+  ///
+  /// ![Butt Cap Diagram](https://flutter.github.io/assets-for-api-docs/assets/dart-ui/butt_cap.png)
+  ///
+  /// The cap is zero-width here, so there is nothing to highlight. Compare to
```
Comment (Hixie) [inline lib/ui/painting.dart]: For someone who is reading this out of context, without seeing the other enum values, this doesn't really make sense. What does "highlight' mean in this context? I'd just drop the sentence entirely.

File: lib/ui/painting.dart (hunk 3 above)
Comment (Hixie) [reply lib/ui/painting.dart]: I'd say something like "In the diagram above", maybe.

File: lib/ui/painting.dart [hunk 5]
```diff
@@ -824,37 +824,73 @@ enum FilterQuality {
...
+/// * [Paint.strokeJoin] and [Paint.strokeMiterLimit] for how this value is
+///   used.
+/// * [StrokeCap] for the different kinds of line endings.
 // These enum values must be kept in sync with SkPaint::Join.
 enum StrokeJoin {
   /// Joins between line segments form sharp corners.
+  ///
+  /// {@animation joinMiter 300 300 https://flutter.github.io/assets-for-api-docs/assets/dart-ui/miter_4_join.mp4}
+  ///
+  /// See also: [Paint.strokeJoin] and [Paint.strokeMiterLimit].
```
Comment (Hixie) [inline lib/ui/painting.dart]: use the bullet style, so you can explain what those are

File: lib/ui/painting.dart [hunk 6]
```diff
@@ -824,37 +824,73 @@ enum FilterQuality {
...
+  /// See also: [Paint.strokeJoin].
   round,
 
   /// Joins between line segments connect the corners of the butt ends of the
   /// line segments to give a beveled appearance.
+  ///
+  /// {@animation joinBevel 300 300 https://flutter.github.io/assets-for-api-docs/assets/dart-ui/bevel_join.mp4}
+  ///
+  /// See also: [Paint.strokeJoin].
   bevel,
```
Comment (Hixie) [inline lib/ui/painting.dart]: ditto with these

File: lib/ui/painting.dart [hunk 7]
```diff
@@ -1055,8 +1091,24 @@ class Paint {
...
+  /// Some examples of joins:
+  ///
+  /// {@animation joinMiter 300 300 https://flutter.github.io/assets-for-api-docs/assets/dart-ui/miter_4_join.mp4}
+  ///
+  /// {@animation joinRound 300 300 https://flutter.github.io/assets-for-api-docs/assets/dart-ui/round_join.mp4}
+  ///
+  /// {@animation joinBevel 300 300 https://flutter.github.io/assets-for-api-docs/assets/dart-ui/bevel_join.mp4}
+  ///
+  /// The centers of the line segments are colored here to highlight the joins,
+  /// but in normal usage the join is the same color as the line.
```
Comment (Hixie) [inline lib/ui/painting.dart]: as above

File: lib/ui/painting.dart [hunk 8]
```diff
@@ -1079,6 +1131,15 @@ class Paint {
...
   /// Defaults to 4.0.  Using zero as a limit will cause a [StrokeJoin.bevel]
   /// join to be used all the time.
+  ///
+  /// {@animation joinMiter 300 300 https://flutter.github.io/assets-for-api-docs/assets/dart-ui/miter_0_join.mp4}
+  ///
+  /// {@animation joinMiter 300 300 https://flutter.github.io/assets-for-api-docs/assets/dart-ui/miter_4_join.mp4}
+  ///
+  /// {@animation joinMiter 300 300 https://flutter.github.io/assets-for-api-docs/assets/dart-ui/miter_6_join.mp4}
+  ///
+  /// The centers of the line segments are colored here to highlight the joins,
```
Comment (Hixie) [inline lib/ui/painting.dart]: ditto

**Comment by Hixie** [comment]
LGTM.

The only change I would make would be to turn the alternative text into actual paragraphs that can be used as true alternatives to the diagrams. Right now they're more like sentence fragments and it's not clear what they mean. Consider how it would read:

> StrokeCap.butt. Begin and end contours with a flat edge and no extension. Line with a square end that stops at the end of the line. Compare to the [square] cap, which has the same shape, but extends past the end of the line by half a stroke width.

That middle sentence doesn't really make much sense in that context. To write alternative text, you have to imagine what you would want the image to be replaced by when someone was reading the documentation to you over the phone. It could be a multi-sentence-long paragraph of information, it could be nothing, it depends on what _additional_ information the diagram is conveying. (If the diagram merely provides an alternative way to explain what the prose already says, for example, then the best alternative text is nothing.)

**Comment by gspencergoog** [comment]
OK, I updated those alt text to be complete sentences with substance.  I was thinking of the alt text as a simple description of the diagram contents (a la the descriptive closed captioning track on a movie), not as replacement prose. It does make more sense to do it as you describe it.

**Pull request merged**